# Learning Objectives

- [ ] 2.5.1 Define and understand classes and objects.
- [ ] 2.5.2 Understand encapsulation and how classes support information hiding and implementation independence.
- [ ] 2.5.3 Understand inheritance and how it promotes software reuse. 
- [ ] 2.5.4 Understand polymorphism and how it enables code generalisation. *Exclude: method overloading and multiple inheritance*

# References

1. Garavaglia, Emilio. “What was the motivation of object oriented programming?” Quora, 30 Mar 2018. Quora, https://www.quora.com/What-was-the-motivation-of-object-oriented-programming. Accessed 8 May 2020.
2. 9. Classes¶. (n.d.). Retrieved May 9, 2020, from https://docs.python.org/2/tutorial/classes.html
3. Encapsulation in Python - GeeksforGeeks. (2019, October 15). Retrieved May 9, 2020, from GeeksforGeeks website: https://www.geeksforgeeks.org/encapsulation-in-python/
4. Polymorphism in Python(with Examples). (2020). Retrieved May 9, 2020, from Programiz.com website: https://www.programiz.com/python-programming/polymorphism
5. Lecture 8: Object Oriented Programming | Lecture Videos | Introduction to Computer Science and Programming in Python | Electrical Engineering and Computer Science | MIT OpenCourseWare. (2016). Retrieved May 9, 2020, from Mit.edu website: https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-0001-introduction-to-computer-science-and-programming-in-python-fall-2016/lecture-videos/lecture-8-object-oriented-programming/
6. ProgrammingKnowledge. (2020). Python Tutorial for Beginners 27 - Python Encapsulation [YouTube Video]. Retrieved from https://www.youtube.com/watch?v=TFLo9m0jFEg
7. Zhang, Qinjie. (2020) Computing Teacher Upgrading Course.
8. UML Class Diagram. https://www.visual-paradigm.com/guide/uml-unified-modeling-language/uml-class-diagram-tutorial/
9. The @property Decorator in Python: Its Use Cases, Advantages, and Syntax. https://www.freecodecamp.org/news/python-property-decorator/#:~:text=The%20%40property%20is%20a%20built,of%20the%20use%20of%20%40property!

## 9.0. Motivation

Imagine that you have a game idea that you want to implement. For example, a platformer. You would have started with a world for this game. However, at the start, this world be empty,
<center> 
<img src="images/mario_0.jpg" width="250" align="center"/>
</center>
and your code probably look something as such:

>```python
>def CreateWorld():
>   ##Some code here.
>```

To make the 'game' into something playable, you would want to have a player character and an enemy. 

<center>
<img src="images/mario_1.jpg" width="250" align="center"/>
</center>

and your code would be a little bit longer but definitely still looks neat and readable. 

>```python
>def CreateWorld():
>   ##Some code here.
>
>def PlayerChar():
>   ##Some code here.
>    
>def EnemyChar():
>   "##Some code here. 

As you would want your player to **play the game**, you need to define some controls allowed for the player. Also, you would also code in some behaviour of the enemy, probably moving left and right. At this juncture, your code would have bloat a little bit, definitely still managable. 

>```python
>def CreateWorld():
>   ##Some code here.
>
>def PlayerChar():
>   ##Some code here.
>   def Jump():
>      ##Some code here.
>    
>def EnemyChar():
>   ##Some code here.
>   def Move():
>       ##Some code here. 

Now you start to get creative and probably want to have more enemies on screen, different types of enemies, other interactable elements etc, as shown in the figure below. 

<center>
<img src="images/mario_2.jpg" width="250" align="center"/>
</center>

>```python
>def CreateWorld():
>   ##Some code here.
>
>def PlayerChar():
>   ##Some code here.
>    
>def EnemyChar1():
>   ##Some code here.
>
>def EnemyChar2():
>   ##Some code here.
>
>def EnemyChar3():
>   ##Some code here.
>    
>def EnemyChar4():
>   ##Some code here.
>```

You persevered through and as your imagination goes even wilder.

<center>
<img src="images/mario_3.jpg" width="500" align="center"/>
</center>

As your code gets longer and your environment gets more and more complex, you realize that it's getting harder and harder to keep track of what's going on and your code will look like a complete mess. While you can still work on it, everytime you need something, you have to always search it in the middle of that “everything”. This will soon become inefficient and prone to errors.

Here’s where you start to think that some organization is needed:
- Enemy characters can be grouped by their type. (e.g. Koopas and Goombas)
- Interactable elements by their ability to generate coins, 
- Background elements by colors 
- etc. 

After organizing, you can start to probably add more types of elements, like timer and score counter, into the game. Also, add more behaviour to the enemies etc., without disrupting the other groups.

When you stop thinking about your objects as merely individual parts and start thinking about them in terms of the relation that exist between them and organize them based on those relation, you are moving from the simple procedural to object oriented programming.

From this analogy, we see that when the complexity of the software increases, and thus, cannot be anymore be efficiently remembered as a whole. It is natural to start to modularize and to think in term of relations among different things you can group based on their affinity. The immediate next step is defining objects and methods.

## 9.1. Introduction to Object Oriented Programming (OOP)

### 9.1.1 Procedural Programming

We used to group blocks of statements together into ``functions``. We call these functions in sequential order. Such way of structuring a program is called Procedural Programming.

Generally states of the program is defined outside of the function as variables.

Such programming paradigm works well for small programs, which is easy to understand and maintain.

Consider the following code on how to find an area of a rectangle. 

We pass the **dimensions of the rectangle** straight into the ``rectangle_area`` function.

In [5]:
# Example 1.1.1 Procedural Programming

# width=input('input_width')
# length=input('input_length')

# print(width*length)

def rectangle_area(width,height):
    return width*height

w=10
l=20
area=rectangle_area(w,l)
print(area)

200


### 9.1.2 Object Oriented Programming (OOP)

**Object Oriented Programming** is another paradigm which bundles properties and behaviors(methods/functions) together into an ``object`` before making use of them, i.e. An object is a self-contained component which consists of methods and properties to make a particular type of data useful.

Consider the following code on how to find an area of a rectangle and compare it with the previous code. Don't worry too much on the details first but rather, note that in OOP, things of interest are organized into an object first before being used.

1. We first create ``Rectangle`` class,
2. define ``get_area()`` function, which returns area a property of a rectangle,
3. Create a ``Rectangle`` object ``r`` with dimensions 10 and 20
4. Get the area from ``r`` by ``get_area`` function

In [45]:
# Example 1.2.1 Object Oriented Programming
class Rectangle:                                    #Create Rectangle class,
    
    def __init__(self,width,height):                #initializer method
        self.width=width
        self.height=height

    def get_area(self):                             #function that returns a property of a rectangle
        return self.width*self.height               

    def __str__(self):
        return f'Rectangle of width {self.width} and length {self.height}'

r=Rectangle(10,20)                                  #Create a Rectangle object with dimensions 10 and 20
area=r.get_area()                                   #Get the area from the Rectangle object r
print(dir(r))
print(r.width)
print(r.height)

['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'get_area', 'height', 'width']
10
20


In general, when we are doing OOP, we take the following steps:

1. Implement a new object type with a class
    - **define** the class (line 2)
    - define **attributes** (WHAT IS the object) (line 4-6)
    - define **functions/methods** (HOW TO use the object) (line 4 and 8)
2. Use the new object type in code
    - creating **instances** of the object type (line 11 in example above)
    - do **operations** with them (line 12)

In [6]:
x=[] #instance of a class List
x.append('a') #List method
print(x)

['a']


In [20]:
import math

x=math.pi
y=math.sqrt(2)
print(x)
print(y)
dir(x)

3.141592653589793
1.4142135623730951


['__abs__',
 '__add__',
 '__bool__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__float__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getformat__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__int__',
 '__le__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__pos__',
 '__pow__',
 '__radd__',
 '__rdivmod__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rfloordiv__',
 '__rmod__',
 '__rmul__',
 '__round__',
 '__rpow__',
 '__rsub__',
 '__rtruediv__',
 '__set_format__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__truediv__',
 '__trunc__',
 'as_integer_ratio',
 'conjugate',
 'fromhex',
 'hex',
 'imag',
 'is_integer',
 'real']

## 9.2. Classes and Instances

### 9.2.1 Classes

A class is a blueprint for the objects of interest. 
* It contains all the details about such a object, e.g. properties and behaviors.
* To define a class, `class` keyword is used.
* Using our platformer example, one possible class is the class of enemies.
* Some examples of class is Python in-built classes `List`, `Integer` etc, which you would have encountered before.
* A function that is available to the class to define the behaviour of its objects is called a *method*, e.g.
    * `append()` method `List` objects,
    * `join()` method for `str` objects,
    * et cetera.

#### 9.2.1.1 Constructing a Class  

- the ``__init__`` method
    - this is a special method in Python, called a **initializer** method
    - its function is to **initiliaze attributes of the objects in the class**,
- The ``self`` parameter is a variable refers to the object itself. By convention, it is always given the name ``self``.
    - Parameter `self` must be the 1st parameter for any method in the class. The `__init__` method takes in a `self` parameter too.
    - Using `self`, you can access other attributes in the object. 
    - To access other attribute, the syntax is `self.insert_attribute_name_of_the_object_here`

In Python, the minimum template to make a class is:

>```Python
>class <CLASS_NAME>:
>
>   def __init__(self,<ATTRIBUTE_1>):
>       self.<ATTRIBUTE_1>=<ATTRIBUTE_1>
>
>```

In the following example, we will create a class called `Enemy` where each of the objects in the class has 2 attributes, `name` and `hp`, which stands for hit points.

In [28]:
# Example 2.1.1 Constructing a Class
class Enemy:
    def __init__(self,name,hit_points): #__init__.() Note that self is passed first
        self.name=name                  
        self.hp=hit_points              #Note that we're using hp instead of hit_points as the latter is too long

In [30]:
# Example 2.1.1 Constructing a Class
class Frenemy:
    def __init__(notconvention,name,hit_points): #__init__.() Note that self is passed first
        notconvention.name=name                  
        notconvention.hp=hit_points              #Note that we're using hp instead of hit_points as the latter is too long

In [32]:
y=Frenemy('Bowser',2000)
print(y)

#### 9.2.1.2 Creating Instance of a Class
After defining a class, we can then create objects of the class (also called **instance** of the class). This process is called *class instantiation* and uses function notation.  

Continuing with our example of a platformer. Let us instantiate an enemy named Goomba and 1 hit point.

In [29]:
# Example 2.1.2 Constructing a Class
x=Enemy('Goomba',1)     #An instance/object of class Enemy named Goomba and 1 hit point is assigned to x
print(x)

#### 9.2.1.3 Using `__str__` Method to Describe Class Instances
Note that when we try to print the object `x` that we created above, the printout is not very useful other as we can't see what are the attributes of this instance.

- the ``__str__`` method
    - this is another special method in Python, 
    - its function is to **return the string that will be produced when** `print` is called on the instance of the class

Thus, to improve our existing class definition, we add the `__str__` method into the class definition.

In [22]:
# Example 2.1.3 Describing a Class Instance
class Enemy:

    def __init__(self,name,hit_points): 
        self.name=name                  
        self.hp=hit_points              

    def __str__(self):                                                      #Note that self is again passed as an argument into the method
        return f'Enemy({self.name},{self.hp})'                     #To access the name and hp of the instance, we need to use self.name and self.hp
        #return 'The enemy is a {} with {} hp'.format(self.name,self.hp)    #This one gives a description in prose form 'The enemy is a Goomba with 1 hp'

    def __repr__(self):

x=Enemy('Goomba',1)
y=Enemy('Covid',5)     
print(x)
print(y)

Enemy(Goomba,1)
Enemy(Covid,5)


### 9.2.2 Instance and Class Attributes 

Instance and Class Attributes are also called Instance and Class variables respectively.   

#### 9.2.2.1 Instance Attributes 

Instance attributes are owned by each individual object/instance of the class. In this case, each object has its own copy of the attributes.
Instance variables are not shared among instances although they have same name.

In Example 9.2.1.3, the `name` and `hp` variables contains different value in `x` and `y`.

To access the instance attribute, we use the syntax 

>```python
> <object>.<insert_attribute_name_of_the_object_here>
>```

In [38]:
# Example 2.2.1 Instance Attributes
print(x.name,' ',x.hp) 
print(y.name,' ',y.hp)

Goomba   1
Covid   5


#### 9.2.2.2 Class Attributes

Class attributes are shared by all instances of that class.
* There is only one copy of the class variable
* When any one object makes a change to a class variable, all the other instances all see the change
* Class variables are declared before the initiliazer `__init__`

Suppose we want each instance of enemy kill to reward an item, we make the following changes to our class definition.

In [39]:
# Example 2.2.2 Describing a Class Instance
class Enemy:

    loot='potato'                          #Class Attribute declared before __init__

    def __init__(self,name,hit_points): 
        self.name=name                  
        self.hp=hit_points              

    def __str__(self):                                                     
        return 'Enemy({},{})'.format(self.name,self.hp)                     

x=Enemy('Goomba',1)
y=Enemy('Goomba',5)

print(Enemy.loot)   
print(x.loot)       #You can access the class attribute at the instance level
print(y.loot)
print(Enemy.hp)     #You cannot access the instance attribute at the class level

potato
potato
potato


AttributeError: type object 'Enemy' has no attribute 'hp'

### 9.2.3 Methods

As mentioned earlier, methods are functions which are added to a class to define the behaviors of the instances of the class. 

There are 3 types of methods that you can include inside a class

* Instance Methods,
* Class Methods, and
* Static Methods.

### 9.2.3.1. Instance Methods

- Instance methods can freely access attributes and other methods on the same object.
- It needs to take in `self` as a parameter to access the attributes of the instance.

In our example, besides having a name and hit point, we would also want the enemy to behave in a certain way. For example, 

1. When the enemy is hit, its hit point get reduced by 1, 
2. When its hit point is equal to 0, it says a final message before it dies,
3. After it dies, it drops an item from its loot table, which is shared across all enemy of the same class.

We will continue our example code and incorporate `get_hit` method  inside the class definition.

In [40]:
# Example 2.3.1 Instance Method
class Enemy:

    loot='potato'                                                       #Class Attribute

    def __init__(self,name,hit_points): 
        self.name=name                  
        self.hp=hit_points              

    def __str__(self):                                                      
        return 'Enemy({},{})'.format(self.name,self.hp)                     
     
    def get_hit(self):                                                  #get_hit is an INSTANCE METHOD as you need to keep track of the hp of the enemy
        self.hp=self.hp-1                                               #1 HP decrease after each hit         
        if self.hp==0:
            print('Oh nuuuuuu.....Farewell, cruel world! (T.T)')        #The dying message
            print('Congratulations. You get {}.'.format(self.loot))     #Loot dropped
            return 
        print('Still strong with {} HP left!'.format(self.hp))          #Taunt yells
        return
    
x=Enemy('Goomba',3)
x.get_hit()
x.get_hit()
x.get_hit()

Still strong with 2 HP left!
Still strong with 1 HP left!
Oh nuuuuuu.....Farewell, cruel world! (T.T)
Congratulations. You get potato.


### 9.2.3.2 Class Methods

- Takes in `cls` as a parameter,
- Class methods cannot freely access the attributes of the instantiated objects
- Class methods can only access class attributes
- Before declaring a class method, put `@classmethod` decorator on the line above it 
- Similar to instance method, instead of `self`, class methods have `cls` as the first argument

Suppose that we want to change the loot droppable in the game after the first enemy death, we can change the class variable `loot` by using a class method.
Let us call this class method `drop_loot`.

In [43]:
# Example 2.3.2 Class Method 
class Enemy:

    loot='potato'

    def __init__(self,name,hit_points): 
        self.name=name                  
        self.hp=hit_points              

    def __str__(self):                                                      
        return 'Enemy({},{})'.format(self.name,self.hp)                     
    
    @classmethod                                                        #Class method decorator
    def drop_loot(cls):                                                 #First parameter is cls
        print('Congratulations. You get {}.'.format(cls.loot))
        cls.loot='banana'                                               #After the first potato drop, the loot becomes a banana instead   
  
    def get_hit(self):                                              
        self.hp=self.hp-1                                                   
        if self.hp==0:
            print('Oh nuuuuuu.....Farewell, cruel world! (T.T)')
            self.drop_loot()                                            #Note that the instance still can access class method
            return 
        print('Still strong with {} HP left!'.format(self.hp))         
        return
    
x=Enemy('Goomba',3)
x.get_hit()
print(x.loot)
x.get_hit()
print(x.loot)
x.get_hit()
print(x.loot)

y=Enemy('Goomba',2)
y.get_hit()
print(y.loot)
y.get_hit()

z=Enemy('Goomba',1)
z.get_hit()

Still strong with 2 HP left!
potato
Still strong with 1 HP left!
potato
Oh nuuuuuu.....Farewell, cruel world! (T.T)
Congratulations. You get potato.
banana
Still strong with 1 HP left!
banana
Oh nuuuuuu.....Farewell, cruel world! (T.T)
Congratulations. You get banana.
Oh nuuuuuu.....Farewell, cruel world! (T.T)
Congratulations. You get banana.


### 9.2.3.3 Static Methods
    
- Takes neither a `self` or `cls` as parameter,
- Cannot modify attributes of class nor instances.
- Before declaring a static method, put `@staticmethod` decorator on the line above it
- Can be used when the method doesn't depend on the class nor instance attributes.

In our example, we see that we could code to print the dying message doesn't rely on any attributes of the instance nor the class. Let's name the static method `death_cry`.

<center>

| Enemy             |
|------------------------|
|------------------------|
| loot: STRING        |
| name: STRING        |
| hp: INTEGER        |
|------------------------|
| constructor(n:name,h:hit_points): |
| death_cry(): |
| drop_loot(c:Enemy): |
| get_hit():|

</center>

In [44]:
# Example 2.3.3 Static Method 
class Enemy:

    loot='potato'

    def __init__(self,name,hit_points): 
        self.name=name                  
        self.hp=hit_points              

    def __str__(self):                                                      
        return 'Enemy({},{})'.format(self.name,self.hp)   

    @staticmethod                                                       #Static Method Decorator
    def death_cry():                                                    #No self nor cls as parameter
        print('Oh nuuuuuu.....Farewell, cruel world! (T.T)')                  
    
    @classmethod                                                       
    def drop_loot(cls):                                                
        print('Congratulations. You get {}.'.format(cls.loot))
        cls.loot='banana'                                                  
  
    def get_hit(self):                                              
        self.hp=self.hp-1                                                   
        if self.hp==0:
            self.drop_loot()
            self.death_cry()                                            #Note that you still need self.death_cry()
            return 
        print('Still strong with {} HP left!'.format(self.hp))         
   
    
x=Enemy('Goomba',3)
x.get_hit()
x.get_hit()
x.get_hit()

y=Enemy('Goomba',2)
y.get_hit()
y.get_hit()

z=Enemy('Goomba',1)
z.get_hit()

Still strong with 2 HP left!
Still strong with 1 HP left!
Congratulations. You get potato.
Oh nuuuuuu.....Farewell, cruel world! (T.T)
Still strong with 1 HP left!
Congratulations. You get banana.
Oh nuuuuuu.....Farewell, cruel world! (T.T)
Congratulations. You get banana.
Oh nuuuuuu.....Farewell, cruel world! (T.T)


### Assignment 1

### a)

Create a class `Person` which contains 2 properties, `_name` and `_age`. 
* Implement its initializer method which initialize `_name` and `_age`.
* Implement its `__str__()` method which return string in the format of `"Person: name={_name}, age={_age}"`
* Define a class variable `MIN_ADULT_AGE` with value `18`. 
* Implement an instance method `is_adult()` which returns `True` if the age is equal or above `MIN_ADULT_AGE`, else returns `False`.

**Sample Output:**
```
Person: name=Alan, age=20
Is adult: True
```

In [ ]:
class Person:
    pass
    #YOUR CODE HERE

# TEST
p = Person('Alan', 20)
print(p)
print('Is adult:', p.is_adult())

Person: name=Alan, age=20
Is adult: True


### b) 

Create a class `Shape` which contains following attributes:
* Define a class variable `PI` with value `3.14`.
* Define a class method `area_circle()` which takes in a `radius` value and returns area of the circle.
* Define a static method `area_rectangle()` which takes in `width` and `length` values, and returns area of the rectangle. 

**Sample Output:**
```
Area of Circle: 12.56
Area of Rectangle: 8
```

In [ ]:
class Shape:
    pass
    #YOUR CODE HERE

# TEST
print('Area of Rectangle:', Shape.area_rectangle(2,4))
print('Area of Circle:', Shape.area_circle(2))

Area of Rectangle: 8
Area of Circle: 12.56


## 9.3 Unified Modeling Language (UML) Class Diagrams

A class diagram is a table that is used to represent classes and it is broken down to 3 partitions:
- Class Name appears in the first partition and is the only mandatory information
- Class Attributes are located in the 2nd partition and types of the attribute is shown after the colon
- Class Methods are shown in the 3rd partition. Return type of method parameters are shown after colon following the parameter name.

<center>

| CLASS_NAME             |
|------------------------|
|------------------------|
| ATTRIBUTE: TYPE        |
|------------------------|
| METHOD(ATTRIBUTE: TYPE):TYPE |

</center>


UML is useful to showcase how classes in the program are related to one another.

## 9.3. Encapsulation and Information Hiding

Recall that a class in Python is a way to group some objects together based on their attributes and methods. This idea of grouping is called **encapsulation**. So, put it in another way, a class encapsulates all the data that is member functions, variables/attributes, etc.

<center>
<img src="images/encapsulation.jpg" width="250" align="center"/>
</center>

* Encapsulation is a way to restrict access to object attributes and methods, 
* Encapsulation is to prevent accidental modification of internal data into possibly sth invalid. see https://www.youtube.com/watch?v=TFLo9m0jFEg
* `set_attr` and `get_attr` method in classes
* `self.a`, `self._b` and `self.__c`
* whenever `__` is used, your data is made private, 
* single underscore `_` is partially private and only a convention.


Class methods have access to all the data in the object.

Example of catastrophic modification of code.
* Possible idea, student average mark over 2 years of A-levels
* average, smallest, largest, largest_count
* corruption of original data.


### 9.3.1 Revisiting Class 

In Section 9.2, when we reintroduced class, we see that a class acts as a template for objects with a certain attributes and methods. This is a useful implementation of classes. However, there is really nothing preventing us from defining a class without attributes and methods. This is the simplest type of class. To check the instance attributes, we use the `__dict__` method, a special method that every object has. It returns a dictionary containing all the object attributes and its associated values.

In [35]:
#Example 3.1.1 Class without attribute nor method
class Enemy:
    pass

x=Enemy()
print(Enemy)                 #An instance of Enemy
print(x.__dict__)            #Printing out the instance variables of x

<class '__main__.Enemy'>
{}


Note that such objects aren't interesting and seems pretty uselees. In Python, values can assigned to a non-existent instance variable of an object and such instance variable will be created immediately.

In [36]:
#Example 3.1.2 Adding instance attributes
x.name='Derpy Derp'         #adding an attribute name to x
x.hp=5                      #adding an attribute hp to x
print(x.__dict__)           

{'name': 'Derpy Derp', 'hp': 5}


Therefore, we see that Python is very flexible in this regard as we can add attributes to objects on the fly. However, this means that the internal properties of the instances can be modified externally and this situation can be undesirable, e.g. some of the attributes get changed accidentally and causes incorrect behaviour when these attributes need to be used.

### 9.3.2 Information Hiding

All methods and variables in a Python class or object are public, i.e. they can be accessed by users. This is the reason why attributes can be modified externally as shown in the subsection 3.1. 

- Python has NO access modifier, i.e. like public & protected & private in C# or Java.
- It uses a convention with a combination of single underscore `_` and double underscore `__` in the naming to indicate whether a method or attribute is for system use or class-internal use.
- Such methods and attributes **should** not be used directly by users of the class. But you can still access them directly, which is useful for debugging purpose.

Before we go into the naming convention, we see the implementation first.

In [41]:
#Example 3.2.1 Information Hiding
class Enemy:
    def __init__(self,name,hp,weapon):
        self.name=name
        self._hp=hp                 #starts with single underscore as the instance variable name _hp
        self.__weapon=weapon        #starts with double underscore as the instance variable name __weapon

x=Enemy('Derpy Derp',5,'axe')
print(x.name)           #Attempting to access name
print(x._hp)            #Attempting to access hp,  usually we do x.hp
#print(x.__weapon)       #Attempting to access weapon
print(x.__dict__)

Derpy Derp
5
{'name': 'Derpy Derp', '_hp': 5, '_Enemy__weapon': 'axe'}


### 9.3.2.1 Naming Conventions for Information Hiding
 
The following are the naming convention for the attributes and methods:

### a) System Attribute `__attr__`

- Attributes with **double-leading and double-trailing underscores** are defined by Python. They are called `magic attributes` or `system attributes`. Such attributes should not be used. 
- For example, the `__class__`, `__name__` property, the `__init__()` and `__str__()` methods.

### b) Class/Module Attribute `_attr`

- Attributes with **single-leading underscores** are for internal use in the class or module. 
- This is just a **convention** which has no effect to Python interpreter. In Example 3.2.1, `x._hp` still returns a value.
- **Note:** When a module is imported, method and variable with single-leading-underscore will NOT be imported.

### c) Name Mangling Attribute `__attr`

- In Example 3.2.1, `x._weapon` causes an `AttributeError`.  
- The reasons is that when a class attribute is defined with **double-leading-underscore**, Python interpretor will prefix such attributes with `_classname`, e.g. `__weapon` in class `Enemy` will become `_Enemy__weapon`. 
- This process of prefixing of `_classname` to the attributes is called **name mangling**.

In [42]:
#Example 3.2.2 Name Mangling
class Enemy:
    def __init__(self,name,hp,weapon):
        self.name=name
        self._hp=hp                
        self.__weapon=weapon        #This attribute is going to be name mangled

    def _death_cry(self):
        pass

    def __drop_loot(self):
        pass

x=Enemy('Derpy Derp',5,'axe')
print(x.__dict__)                    #Checking the instance variables of instance x
print(Enemy.__dict__)                #Checking the method of object Enemy

{'name': 'Derpy Derp', '_hp': 5, '_Enemy__weapon': 'axe'}
{'__module__': '__main__', '__init__': <function Enemy.__init__ at 0x0000024E0924A670>, '_death_cry': <function Enemy._death_cry at 0x0000024E0924AC10>, '_Enemy__drop_loot': <function Enemy.__drop_loot at 0x0000024E0924A0D0>, '__dict__': <attribute '__dict__' of 'Enemy' objects>, '__weakref__': <attribute '__weakref__' of 'Enemy' objects>, '__doc__': None}


In conclusion, to hide an attribute from external access, we should use double-leading-underscore to prefix the variable name `__`.

### 9.3.2.2 Retrieving Information

In the previous subsection, we have seen how to shield the outside eyes from peeking into the attributes of the instance. However, there are instances where you would want to allow the instance attribute to be able to be retrieved and modified by an external code. For example, you might want to create a graphical intercace to display the enemy's hp etc.

To do this, there are two possible ways to allow this:
- `getter` and `setter` method
- `@property` decorator and `@attr.setter` decorator

### a) `getter` and `setter` method

- A `getter` is a method that gets the value of an attribute and similarly, a `setter` is a method that sets the value of a property.
- These methods usually come in pair for each attribute, but not necessarily.

In [50]:
#Example 3.2.3 getter and setter method approach
class Enemy:
    
    def __init__(self,name,hp,weapon):
        self._name=name
        self.__hp=hp                  
    
    def get_name(self):                     #getter for name
        return self._name
    
    def set_name(self,val):                 #setter for name, note that it has another parameter val that is used to set the instance variable _name
        self._name=val

    def get_hp(self):                       #getter for hp
        return self.__hp                    #Research into this

    def set_hp(self,val):                   #setter for name, note that it has another parameter val that is used to set the instance variable __hp
        self.__hp=val

x = Enemy('Koopa',4,'shell')

x.name='lel'
print('1st get_name() call: name of x is',x.get_name()) #Note that the instance variable doesn't get changed
x.set_name('lel')
print('2nd get_name() call: name of x is',x.get_name())

print('1st get_hp() call: hp of x is',x.get_hp())
x.set_hp(5)
print('2nd get_hp() call: hp of x is',x.get_hp())

x.__dict__

1st get_name() call: name of x is Koopa
2nd get_name() call: name of x is lel
1st get_hp() call: hp of x is 4
2nd get_hp() call: hp of x is 5


{'_name': 'lel', '_Enemy__hp': 5, 'name': 'lel'}

### b) `@property` decorator
An alternative way is to use `@property` decorator. 
* The `@property` decorator marks the getter method
* The `@attr.setter` decorator marks the setter method for attribute `attr`

Using this decorators makes the attribute into a `property` object, which could then be used in a more natural manner.

It is common to use `@property` to implement a read-only computed attribute, which doesn't require taking in of parameters, e.g., `name` and `hp` in the `Enemy` class.

In [51]:
#Example 3.2.4 @property and @attr.setter approach
class Enemy:
    
    def __init__(self,name,hp,weapon):
        self._name=name
        self.__hp=hp
        self._weapon=weapon                
    
    @property                           #marks the getter method
    def name(self):                     
        return self._name
    
    @name.setter                        #marks the setter method
    def name(self,val):                 
        self._name=val

    @property                           #marks the getter method
    def hp(self):                      
        return self.__hp

    @hp.setter                          #marks the getter method
    def hp(self,val):                  
        self.__hp=val

    @property
    def weapon(self):
        return self._weapon

    @weapon.setter
    def weapon(self,val):
        self._weapon=val

x = Enemy('Koopa',4,'shell')

print('1st call: name of x is',x.name) #note the syntax used here that makes it natural x.name instead of x.get_name()
x.name='lel'                                      #note the syntax used here that makes it natural x.hp instead of x.hp()
print('2nd call: name of x is',x.name)

print('1st call: hp of x is',x.hp)
x.hp=5
print('2nd call: hp of x is',x.hp)

print(x._name)
print(x._Enemy__hp)
print(x.name)


1st call: name of x is Koopa


### Assignment 2

Create a class `Celsius` with following attributes:
* It contains an instance variable `_value`, which contains its temperature value.
* Implement its `__init__()` method which initialize its instance variable `_value`.
* implement a property `value` for its variable `_value`. 
* implement a read-only property `fahrenheit` which returns temperature value in Fahrenheit.
    * `Fahrenheit = Celsius * 1.8 + 32`

**Sample Output:**
```
Temperature: 25 Celsius, 77.0 Fahrenheit
```

In [ ]:
class Celcius:
    pass
    #YOUR CODE HERE

#TEST
x=Celcius(25)
print(f'Temperature: {x.get_celsius()} Celsius, {x.get_fahrenheit()} Fahrenheit')

Temperature: 25 Celsius, 77.0 Fahrenheit


## 9.3 Unified Modeling Language (UML) Class Diagrams part II

<center>

| `Complex`                      |
|---------------------------------------|
|---------------------------------------|
|  `real: FLOAT`               |
|  `imag: FLOAT`               |
|---------------------------------------|
|  `constructor(real: FLOAT, imag: FLOAT)`         |
|  `Re(): Complex`          |
|  `Im(): Complex`            |
|  `mod(): FLOAT`            |
|  `arg(): FLOAT`            |
|---------------------------------------|

</center>



<center>

| `ToDo`                      |
|------------------------------------------|
|------------------------------------------|
|  `category: STRING`               |
|  `description: STRING`               |
|------------------------------------------|
|  `constructor(c : STRING, d : STRING)`         |
|  `set_category(s : STRING)`          |
|  `set_description(s: STRING)`            |
|  `get_category(): STRING`       |
|  `get_description(): STRING`  |
|  `summary(): STRING` |
|------------------------------------------|

</center>

| `__str__(): STRING`|	Returns the string representation of the complex number in the form `<real_part>+i*<imaginary_part>`|

## 9.4. Inheritance and Polymorphism

**Inheritance** is the ability of a class (the derived class) to use the properties and methods of another class (the parent class). This is one of the major benefits of object oriented programming. 

* **Parent Class** is the class being inherited from, also called **base class**.
* **Child class** is the class that inherits from another class, also called **derived class** or a **subclass**

**Benefits:**

* <u>Reuse Quality Code:</u> Reuse existing code which is already tested.
* <u>Improve Code Readability:</u> Program structure is short and concise.
* <u>Improve Code Reliability:</u> Avoid code duplication and easier to debug.
* <u>Save Time and Effort</u> 

In video games, some enemies are variants of a simpler version of itself. For example, here are some Hammer Bros variant from Mario Paper. The enemies at the bottom is derived from the base model at the top. The base model represents the Parent class and the variants the Child Class. 

<center>
<img src="images/hammer_bros_edited.jpg" width="400" align="center"/>
</center>

### Some Basic Syntax

* Without specifying parent class, the class inherits from `object` class.
* The `__base__` attribute of a class returns its base class.
* `issubclass()` function checks whether a class is a subclass of another.
* `__class__` is a special Python method of instances of the class used to get the class, 
* `__name__` is another special Python class method to get the name of the class.

In [55]:
#Basic Syntax to create a Subclass
class Parent:
    pass

class Child(Parent): #This is the syntax to create a Child class out of Parent class
    pass

In [58]:
Parent.__base__
Child.__base__
Child.__name__

'Child'

### 4.1 Inheritance

To illustrate how inheritance work, we start by defining a base class `HammerBro`, which has property `base_damage`, `hp` and `damage_mod`. It also has a method `attack_damage()` which calculates the attack damage it can inflict on the player character and `attack()` that prints 'HammerBro attacks for 1 damage.'. We will: 
* define the class attribute `base_damage`, 
* Initialize its property `hp` and `damage_mod` in its constructor function, i.e. `__init__()` function,
* Implement its `__str__()` function which returns string `HammerBro has 5 HP and wields hammer` on the instantiated object that describes the object,
* define the method `attack_damage` and `attack`,

In [64]:
#Example 4.1.1 Creation of a Parent Class
class HammerBro:

    base_damage=1

    def __init__(self,hp=5,damage_mod=1):
        self._hp=hp
        self._damage_mod=damage_mod

    @property
    def hp(self):
        return self._hp

    @property
    def damage_mod(self):
        return self._damage_mod

    def __str__(self):
        return '{} has {} HP and wields hammer.'.format(self.__class__.__name__,self._hp) #Note the use of __class__.__name__ here to get the name of the class

    def attack_damage(self):
        return self.base_damage*self.damage_mod

    def attack(self):
        return print(self.__class__.__name__,'attacks for',self.attack_damage(),'damage')

h=HammerBro()
print(h.base_damage)        #Class variable base_damage is inherited
print(h.hp)                 #Instance variable
print(h.damage_mod)         #Instance variable
print(h)
h.attack()

1
5
1
HammerBro has 5 HP and wields hammer.
HammerBro attacks for 1 damage


Next, we will define a subclass `IceBro` from `HammerBro`. 
* Note that without any further coding, `IceBro` class will be  able to access to all attributes in `HammerBro` class.

In [66]:
#Example 4.1.2 Creation of a Child Class / Subclass

class IceBro(HammerBro):    #Defining the subclass IceBro of HammerBro
    pass
        
x=IceBro()                  #Creating an IceBro Object and assigning to x
print(x.base_damage)        #Class variable base_damage is inherited
print(x.hp)                 #Instance variable
print(x.damage_mod)         #Instance variable
print(x)
x.attack()
print(issubclass(IceBro,HammerBro)) #Checking if IceBro is a subclass of HammerBro

1
5
1
IceBro has 5 HP and wields hammer.
IceBro attacks for 1 damage
True


However, note that when we print the object, it says that "IceBro has 5 HP and wields hammer". This makes IceBro sad because he uses ice to attack. (T.T)

### 9.4.2 Method Overriding

To correct and improve the construction of the subclass objects, we should include the different type of weapons that the variants use.
This means that we should have an instance variable `weapon` to store the type of the weapon, 
Thus, when we are initializing the instance, we need to include this `weapon`.
Consequently, we need to define a new `__init__` function for the subclass,

This ability of a class to change the implementation of a method provided by one of its ancestors is called **method overriding**.

In [67]:
#Example 4.2.1 Method Overriding

class VariantBro(HammerBro):    #Defining the subclass VariantBro of HammerBro
    
    base_damage=2               #We are overriding the base_damage in HammerBro
    
    def __init__(self,weapon,hp=5,damage_mod=1): #We override the __init__ 
        self._hp=hp
        self._damage_mod=damage_mod
        self._weapon=weapon

    @property                   #VariantBro can have more methods than HammerBro
    def weapon(self):
        return self._weapon

    def __str__(self):          #Override __str__ from HammerBro
        return '{} has {} HP and wields {}.'.format(self.__class__.__name__,self._hp,self._weapon)
        
ice=VariantBro(damage_mod=1.1,weapon='ice')             #Creating an VariantBro Object to represent Ice Bro      
fire=VariantBro(hp=7,damage_mod=1.5,weapon='fire')      #Creating an VariantBro Object to represent Fire Bro
bone=VariantBro(hp=10,weapon='bone',damage_mod=0.8)     #Creating an VariantBro Object to represent Dry Bones Bro
boom=VariantBro(weapon='boomerang')                     #Creating an VariantBro Object to represent Boomerang Bro
print(ice)
print(fire)
print(bone)
print(boom)
ice.attack()    
fire.attack()
bone.attack()
boom.attack()

x=HammerBro()
x.attack()

VariantBro has 5 HP and wields ice.
VariantBro has 7 HP and wields fire.
VariantBro has 10 HP and wields bone.
VariantBro has 5 HP and wields boomerang.
VariantBro attacks for 2.2 damage
VariantBro attacks for 3.0 damage
VariantBro attacks for 1.6 damage
VariantBro attacks for 2 damage
HammerBro attacks for 1 damage


Note that after implementation, the call initializer `VariantBro` expects 3 positional argument `weapon`, `hp`, `damage_mod`,  instead of just the first two. 

### 9.4.2.1 The `super()` method

As mentioned in in the previous subsection, the `VariantBro.__init__` method overrides the `HammerBro.__init__` method.

However, in `VariantBro.__init__` method, the first two lines are actually just a repeat of the `HammerBro.__init__` method. 

This example illustrates that sometimes accessing the parent version of overriden attribute(s) could be useful.

The `super()` returns object of parent class when called inside a subclass.

In [155]:
#Example 4.2.2 The super() method
class VariantBro(HammerBro):    
    
    base_damage=2               
    
    def __init__(self,weapon,hp=5,damage_mod=1): #We override the __init__ 
        super().__init__(hp,damage_mod)          #We use super() to access the __init__ in HammerBro
        self._weapon=weapon

    @property                   
    def weapon(self):
        return self._weapon

    def __str__(self):          
        return '{} has {} HP and wields {}.'.format(self.__class__.__name__,self._hp,self._weapon)
        
ice=VariantBro(damage_mod=1.1,weapon='ice') 
print(ice)
ice.attack()    

VariantBro has 5 HP and wields ice.
VariantBro attacks for 2.2 damage.


### 9.4.3 Polymorphism

The ability of a method to exhibit different behaviours depending on the object on which the method is invoked is termed **polymorphism**. In Example 9.44.2.1 above, the methods `__init__` and `__str__` from the `VariantBro` subclass has overriden the methods with similar name in the parent class `HammerBro`.

Another example of polymorphism that you might have encounter before is given in the code block below.

In [71]:
#Example 4.3.1 Polymorphism
num1 = 1
num2 = 2
# print(num1+num2)
print(num1.__add__(num2))

num1 = "Python"
num2 = "Programming"
# print(num1+" "+num2)
print(num1.__add__(num2))

3
PythonProgramming


Here, we can see that a single operator `+` has been used to carry out different operations for distinct data types. This is one of the most simple occurrences of polymorphism in Python.

### Assignment 3

Create a class `Employee` which has a variable, `_name`. 
* Implement its initializer method which initialize `_name`.

Create another class `SalaryEmployee` whose payroll is by monthly salary. It has instance variables `_name`, `_salary`.
* Extend `SalaryEmployee` from `Employee`. 
* Implement its initializer method to initialize both `_name` and `_salary`.
* Implement an instance method `get_payroll()` which returns its `_salary` value.

Create another class `CommissionEmployee` whose payroll is by salary+commission. It has instance variables `_name`, `_salary` and `_commission`.
* Extend `CommissionEmployee` from `SalaryEmployee`.
* Implement its initializer method to initialize `_name`, `_salary` and `_commission`.
    * Make use of `__init__()` method in parent class.
* Implement an instance method `complete_project()` which will increase `_commission` by 100.
* Implement an instance method `get_payroll()` which returns `_salary + _commission`.
    * Make use of `get_payroll()` method in parent class.


**Sample Output:**
```
Payroll: 1200
```

In [ ]:
# YOUR CODE HERE

# TEST
ce = CommissionEmployee('Alan', 1000, 100)
ce.complete_project()

print('Payroll:', ce.get_payroll())

In [4]:
class Point:
    def __init__(self,x,y):
        self.x=x
        self.y=y

a=Point(1,5)
a.x=2
a.x

2